gadgets/trace_open_rust/rust/src/lib.rs

```diff
@@ -0,0 +1,187 @@
+// Copyright 2025 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+use api::rust_bindings::{log::LogLevel,datasources::{DataSource, FieldKind, Field, Data}};
+use std::sync::OnceLock;
+
+static MODE_RAW_F: OnceLock<Field> = OnceLock::new();
+static MODE_F: OnceLock<Field> = OnceLock::new();
+static FLAGS_RAW_F: OnceLock<Field> = OnceLock::new();
+static FLAGS_F: OnceLock<Field> = OnceLock::new();
+
+static FLAG_NAMES: [&str; 13] = [
+    "O_CREAT",
+    "O_EXCL",
+    "O_NOCTTY",
+    "O_TRUNC",
+    "O_APPEND",
+    "O_NONBLOCK",
+    "O_DSYNC",
+    "O_FASYNC",
+    "O_DIRECT",
+    "O_LARGEFILE",
+    "O_DIRECTORY",
+    "O_NOFOLLOW",
+    "O_NOATIME",
+    // Note: O_CLOEXEC would be 1 << 19 (not covered if we're shifting by 6)
+];
+
+fn format_mode(mode: u16) -> String {
+    let file_type = match mode & 0o170000 {
+        0o040000 => 'd', // Directory
+        0o100000 => '-', // Regular file
+        0o120000 => 'l', // Symlink
+        _ => '?',        // Unknown
+    };
+
+    let perms = [
+        (mode & 0o400, 'r'),
+        (mode & 0o200, 'w'),
+        (mode & 0o100, 'x'),
+        (mode & 0o040, 'r'),
+        (mode & 0o020, 'w'),
+        (mode & 0o010, 'x'),
+        (mode & 0o004, 'r'),
+        (mode & 0o002, 'w'),
+        (mode & 0o001, 'x'),
+    ];
+
+    let mut result = String::with_capacity(10);
+    result.push(file_type);
+    for (bit, ch) in perms {
+        result.push(if bit != 0 { ch } else { '-' });
+    }
+    result
+}
+
+fn data_callback(_source: DataSource, data: Data) {
```

Comment:
Do you think it's possible to use a closure capturing the fields instead to avoid using global variables?